text/0000-any_type_self.md

```diff
@@ -0,0 +1,79 @@
+- Feature Name: any_type_self
+- Start Date: 2016-07-26
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow self parameter be any type.
+
+# Motivation
+[motivation]: #motivation
+
+It allows to avoid wrapper structs.
+
+# Detailed design
+[design]: #detailed-design
+
+Currently self parameter type is allowed to be only `Self`, `&Self`, `&mut Self`, `Box<Self>`.
+
+Solution: Allow `self` be any type that depends on `Self`.
```

Comment:
But, how would you construct a value of `Vec<Self>` from a value of `Self`? Clearly, that doesn't make sense.

What you would need is a trait, like `From<Self>` which defines that `self` is constructed from the value in question, but at that point you introduce hidden conversions.

I believe that some form of generalized `AsRef` is what's needed.